PP-750: Add a new node attribute named "partition" to the node object

suresh-thelkar opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Issue-ID
<!--- Replace XXXX below (2 places) with the actual JIRA id -->
* **[PP-750](https://pbspro.atlassian.net/browse/PP-750)**

#### Problem description
* This is one of the tickets of the following EPIC
**[PP-337](https://pbspro.atlassian.net/browse/PP-337)** Multiple schedulers servicing the PBS cluster

This ticket addresses adding new attribute named "partition" to node object.

#### Cause / Analysis
* Adding new node attribute named "partition"

#### Solution description
* Adding new node attribute named "partition"
* Taking care of the following validations as well.

PBS admin/manager can set node's partition attribute to an existing partition name and it's corresponding scheduler will be scheduling jobs on this node.
If nodes are associated to a partition then they can not be linked to any queue which isn't part of that partition. Trying to set a node to a queue which isn't part of it's partition will result into the following error:
```
Qmgr: s n node1 queue=workq1
qmgr obj=stblr3 svr=default: Partition p1 of node is not part of queue
qmgr: Error (15220) returned from server
```

If a node is associated to a queue then trying to set a partition on this node which does not belong to the same queue will result into the following error.

```
Qmgr: s n stblr3 partition=p2
qmgr obj=stblr3 svr=default: Queue q1 of node is not part of partition
qmgr: Error (15219) returned from server

```
If a queue is associated to one or multiple nodes then trying to change partition of this queue to a value other than those that are set on these nodes will result into the following error.
```
Qmgr: s q q1 partition=p2
qmgr obj=q1 svr=default: Invalid partition in queue
qmgr: Error (15221) returned from server
```
We have also written PTL test cases for the same. Also attached the PTL logs to the jira ticket.
https://pbspro.atlassian.net/browse/PP-750

#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [x] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [x] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

bhroam commented on test/tests/interfaces/pbs_node_partition.py:
```diff
@@ -0,0 +1,166 @@
+        :param mgr_cmd: qmgr "MGR_CMD_SET/MGR_CMD_UNSET" cmd,
+        Defaults to MGR_CMD_SET
+        :type mgr_cmd: str
+        :param n_name: name of the node/vnode, Defaults to "hostname of server"
```
Just say vnode.  The docs only refer to vnodes.

suresh-thelkar replied:
Done

bhroam commented on test/tests/interfaces/pbs_node_partition.py:
```diff
@@ -0,0 +1,166 @@
+            msg = ("Error: test_set_node_partition_attr function takes only "
+                   "MGR_CMD_SET/MGR_CMD_UNSET value for mgr_cmd")
+            self.logger.error(msg)
+            exit()
```
Why exit()?  If you want to fail the test do an assertTrue(False, msg)

suresh-thelkar replied:
Done

bhroam commented on test/tests/interfaces/pbs_node_partition.py:
```diff
@@ -0,0 +1,166 @@
+        try:
+            self.server.manager(MGR_CMD_UNSET,
+                                NODE, 'partition',
+                                id=self.host_name, runas=TEST_USER)
```
Add a runas argument to set_node_partition_attr().  It can default to ROOT_USER.

suresh-thelkar replied:
Done

bhroam commented on test/tests/interfaces/pbs_node_partition.py:
```diff
@@ -0,0 +1,166 @@
+        """
+        self.set_node_partition_attr()
+        msg1 = "Unauthorized Request"
+        msg2 = "checking the qmgr error message"
```
This is a failure message.  Maybe make it "didn't receive expected error message" (same with the rest of the test plan).

suresh-thelkar replied:
Done

bhroam commented on test/tests/interfaces/pbs_node_partition.py:
```diff
@@ -0,0 +1,166 @@
+# "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+# trademark licensing policies.
+
+from tests.interfaces import *
```
The PTL guidelines tells us to create larger test plans rather than more smaller test plans.  How about adding these tests to your queue partition test plan?  They both are testing partitions.

suresh-thelkar replied:
We hope you are referring to test suite rather than test plan. For the time being we are adding multisched tag to all the test cases that we write for Multisched feature so that we can make use of this tag to run all the regressions related to Multisched as a whole. Hope this is fine.

bhroam replied:
This goes against the PTL guidance we were given.  What is the problem with creating one larger test suite?  Plus we can't create tags.  I believe the PTL steering committee does that.

bhroam commented on src/lib/Liblog/pbs_messages.c:
```diff
@@ -402,6 +402,9 @@ char *msg_sched_priv_exists = "Another Sched object also has same value for its
 char *msg_cannot_set_route_que = "Route queues are incompatible with the partition attribute";
+char *msg_queue_not_in_partition = "Queue %s of node is not part of partition";
+char *msg_partition_not_in_queue = "Partition %s of node is not part of queue";
+char *msg_invalid_partion_in_queue = "Invalid partition in queue";
```
Are these 3 messages in your EDD?

suresh-thelkar replied:
Yes. EDD is updated with these messages.

suresh-thelkar commented on src/lib/Liblog/pbs_messages.c:
```diff
@@ -402,6 +402,9 @@ char *msg_sched_priv_exists = "Another Sched object also has same value for its
 char *msg_route_que_no_partition = "Can not assign a partition to route queue";
 char *msg_cannot_set_route_que = "Route queues are incompatible with the partition attribute";
+char *msg_queue_not_in_partition = "Queue %s of node is not part of partition";
+char *msg_partition_not_in_queue = "Partition %s of node is not part of queue";
```
Replaced "of node" with "for node" at the end of messages as you suggested.

suresh-thelkar commented on src/server/node_func.c:
```diff
@@ -2907,3 +2912,38 @@ chk_vnode_pool (attribute *new, void *pobj, int actmode)
+	}
+
+	return PBSE_NONE;
+}
```
As you said our current action function architecture does not handle the scenario of one attribute validation depending on another if both are set atomically or at once.  I think one of the following bugs (which says attribute validations are not happening when they are defined through vnodedef file)  that we have raised recently also suffers from this.  We also feel that this is the time for change of architecture for actions functions. 

https://pbspro.atlassian.net/browse/PP-925

bhroam replied:
If I were to guess, I'd suspect the issue with PP-925 is that the action functions are not being called at all.  I didn't suggest fixing this issue now because it would creep the scope of your project.  If you feel now is the right time, I am all for it.  I need it for soft_walltime as well.  I came up with two possibilities.  The first is we pass an additional argument to the action functions which is the entire list of attributes being set.  This means that all the action functions have to do double the work.  They first check if the two attributes are being set together, and if not, they pick one up off the object.  The second idea is to create a copy of the object with all the changes made.  This is problematic because you don't always know what the object is.  There are places where you have function A pass a void* to function B which calls the action function with it.  There is no way for function B to know what type of object it is.  If you have any other ideas, I'd love to hear them.

suresh-thelkar commented on src/server/queue_func.c:
```diff
@@ -425,9 +426,23 @@ queuestart_action(attribute *pattr, void *pobject, int actmode)
 	if (((pbs_queue *)pobj)->qu_qs.qu_type  == QTYPE_RoutePush)
 		return PBSE_ROUTE_QUE_NO_PARTITION;
 
+	for (i=0; i<svr_totnodes; i++) {
```
Done

suresh-thelkar commented on src/server/queue_func.c:
```diff
@@ -425,9 +426,23 @@ queuestart_action(attribute *pattr, void *pobject, int actmode)
+		if (pbsndlist[i]->nd_pque) {
+			if (strcmp(pbsndlist[i]->nd_pque->qu_qs.qu_name, ((pbs_queue *) pobj)->qu_qs.qu_name) == 0) {
+				if ((pbsndlist[i]->nd_attr[ND_ATR_partition].at_flags) & ATR_VFLAG_SET &&
+						(pattr->at_flags) & ATR_VFLAG_SET )
```
Done

bhroam commented:
I'm happy with the changes which have been made so far.  We now just need to fix the action function architecture.  Did you want to do that as part of this checkin or make it a separate one on its own?  I'd lean towards making it one on its own (which you could work on next :)

suresh-thelkar commented:
@bhroam 
I have created the following RFE for action functions architecture change.  Please add to to the ticket If I have missed anything.
https://pbspro.atlassian.net/browse/PP-938

suresh-thelkar commented:
@bhroam 
As you suggested we moved all the partition related PTL test cases under one test suite. Please look into the latest commit.  We also removed the tags. Please provide your further comments/sign off.

bhroam commented on test/tests/interfaces/pbs_partition.py:
```diff
@@ -0,0 +1,297 @@
+        :param start: "started" attribute of queue object, Defaults to "True"
+        :type start: boolean
+        :param n_name: name of the vnode, Defaults to "shortname of server"
+        :type n_name: str
```
I think you can merge q_name and n_name into just name.  You either act upon a queue or a node, but never both.

suresh-thelkar replied:
Done

bhroam commented on test/tests/interfaces/pbs_partition.py:
```diff
@@ -0,0 +1,297 @@
+        msg2 = "checking the qmgr error message"
+        try:
+            self.server.manager(MGR_CMD_SET,
+                                QUEUE, {'queue_type': "r"},
```
spell out routing.  Using the shorthands are great when typing, but I think tests should be more explicit when setting attributes.

suresh-thelkar replied:
Done

bhroam commented on test/tests/interfaces/pbs_partition.py:
```diff
@@ -0,0 +1,297 @@
+                            id="Q1", expect=True)
+        self.server.manager(MGR_CMD_SET,
+                            QUEUE, {'queue_type': "execution"},
+                            id="Q1", expect=True)
```
Why not use partition_attr() to create the above 3 queues?

suresh-thelkar replied:
Done

suresh-thelkar commented:
@bhroam 
We have implemented your review comments. Please look into it and give your further comments if any or sign off.

bhroam commented on test/tests/interfaces/pbs_partition.py:
```diff
@@ -46,10 +46,9 @@ class TestPartition(TestInterfaces):
+                       start="True", partition="P1", user=ROOT_USER):
         """
         Common function to set partition attribute/s to node/queue object
         :param mgr_cmd: qmgr "MGR_CMD_SET/MGR_CMD_UNSET/MGR_CMD_CREATE" cmd,
```
Just a suggestion: Instead of making this a string, how about using the PTL defined symbols that expect() uses (e.g., MGR_CMD_SET)?  That way you're not doing string compares.  It'll be marginally faster.

bhroam commented:
I'm pretty happy with the code as it is.  I just had that one suggestion you can choose to do or not.  We might want to finish having the conversation about setting partition and queue at the same time before we check this in.  You might want to make minor changes based on it.

suresh-thelkar commented:
@bhroam 
We have implemented your optional comment as well and pushed the changes.

bhroam commented:
Looks like travis failed due to a network issue.  I restarted it, so hopefully it'll be successful this time.

bhroam commented:
Please squash and rebase for the final merge.

suresh-thelkar commented:
@bhroam 
I have squashed and rebased it with master.  Please merge this PR.

bhroam commented:
Looks like something got merged after you rebased.  Please rebase again and I'll merge it in.

suresh-thelkar commented:
@bhroam 
I have rebased it again. Can you please do the needful.
